### ISSUE TITLE
Phantom windows

### ISSUE THREAD

**Issue by piotr-dobrogost** [kwin-scripts/kwin-tiling#75]
It seems in some circumstances tiling is loosing track of windows which do not exist anymore. As a result there's free space for such windows left in the layout which should instead be occupied by other, existing windows. Hopefully I will establish what these circumstances are and post more information here.

@faho 
How does tiling handle window lifecycle?

**Comment by faho** [comment]
I had actually assumed I had eliminated all related bugs, damn.

> How does tiling handle window lifecycle?

When a window gets created, kwin sends the "workspace.clientAdded" signal. We listen for that and check if the window is tileable, completely untileable or temporarily untileable (e.g. because it's fullscreen - that can change, so we still connect some signals). If it is, we create a tile for it (or if it's not the current tab, we add it to a tile for its tabgroup, though that's not included in current kwin versions).

When a window is minimized, closed or the user sets it to floating, we remove the tile.

One thing that might look like this is clients that can't actually be tiled (e.g. because they can't be resized) but still have a tile prepared for them. This used to happen with the steam update dialog, which sets window type to normal. 

I had a commit that ignores unresizeable clients, but I had to revert it because that can be a temporary thing.

**Comment by piotr-dobrogost** [comment]
Btw, is there a way to reset state of tiling completely when this happens?

**Comment by faho** [comment]
>There's assumption that we get closed signal when the app crashes, yes?

Yes. Which seems like a fair assumption to make since KWin sends that signal.

>Btw, is there a way to reset state of tiling completely when this happens?

No. The problem is that you'd need to be able to select _tiles_ (not clients), and for that we'd _need_ some way to display graphical stuff, and I haven't figured out how to do that. When I last tried (QML should allow that), I didn't even manage to draw a simple rectangle. It just didn't work.

And if we could figure out which tile to close without the user manually selecting it, we could also just solve the issue.

Anyway, until you figure out when this happens (and I haven't seen it in the past year), there's nothing we can do.

-----

Another possibility though is that you're hitting "the pixmap issue" (which I've reported upstream, and I have some more workarounds in the pipeline). Under some circumstances, windows can get "stuck". Resizing the tile helps with this. Try grabbing a tile and yanking it around a bit.

**Comment by piotr-dobrogost** [comment]
>No. The problem is that you'd need to be able to select tiles (not clients), and for that we'd need some way to display graphical stuff, and I haven't figured out how to do that. When I last tried (QML should allow that), I didn't even manage to draw a simple rectangle. It just didn't work.

Well, you seem to have in mind some fancy repair feature whereas it would be enough to have kind of total reset; delete all tiles and either create them a new and tile all existing windows or just leave it as is without tiling. Btw, how does *Tile now* work?

> And if we could figure out which tile to close without the user manually selecting it, we could also just solve the issue.

Do we have tile→window mapping? If yes then how about iterating through all tiles and checking if the window still exists and deleting tiles without window?

**Comment by faho** [comment]
>Btw, how does Tile now work?

When tiling is deactivated on a desktop, we still keep track of what the geometries should be, we just don't set them. "Tile Now" just sets them once. (Well, what it actually does is toggle the "active" setting twice, but that's an implementation detail that should probably be cleaned up)

----

As for the rest, I think we've gotten off track a bit here. Yes, we could implement a workaround if stuff ever got out of sync. But the more important thing is to fix this bug, and I'm reasonably certain that there's not that many cases where we fail. I haven't seen any in the past year.

So, @piotr-dobrogost: How often does this occur? Is this a once-in-a-blue-moon freak accident or once per boot or every five minutes? Which layout are you using (it's possible the non-defaults, i.e. not "halflayout" have rusted a bit since the hotkeys to switch to them weren't working and that's my default)? What applications are you using when this occurs? Are there any windows involved that just appear for a split second or so?

**Comment by piotr-dobrogost** [comment]
It happened a few times one day and AFAIR it had something to do with Chrome's Developer Tools window which I was switching back and forth to be either embedded in the Chrome's window or to be standalone (detached). Also I'm using multiple monitors and AFAIR I was switching some windows between two screens at the time it happened (using *Window To Next Screen* KWin's keyboard shortcut).

> As for the rest, I think we've gotten off track a bit here. 

It depends on how you look at this. I'm asking about different things which I think are in some way related to the problem. Also, to understand specific features/bugs it's often helpful to have a greater picture and I'm trying to get one by asking these questions. I'm pretty sure people coming here will find information you provide valuable.

> Yes, we could implement a workaround if stuff ever got out of sync.

I was just thinking of some way to reset the state so that people would know how to get out of trouble in case tiling misbehaves in any way. I didn't want to imply we have to create some workaround for this specific case. Is restarting KWin a preferable way of *resetting* tiling script?

> But the more important thing is to fix this bug, and I'm reasonably certain that there's not that many cases where we fail. I haven't seen any in the past year.

Please do not feel obliged in any way to fix something which I'm not yet sure is broken and which I can't reliably reproduce. I created this issue as a place to gather any and all relevant information in case someone encounter such a behavior. Being able to reliably reproduce a bug is essential to start thinking about fixing it and I'm well aware of this.

I'm using halflayout.

**Comment by DaBlitzStein** [comment]
It is broken, it is happening on my system. Latest from kubuntu ppa.

**Comment by TomBZombie** [comment]
I am getting this, but only after adding a second screen. With a single screen the tiling works perfectly. With two, I get phantom windows.

Seems to be sublime text 3 which consistently triggers it for me.

**Comment by TomBZombie** [comment]
Ok, I think I have narrowed down the issue: It's when there are two instances of the same application running but one is minimized. The minimized one is given a tiling slot creating a gap. 

Sublime restores its own state here's how to consistently trigger it:

- Start an application which creates multiple windows (even multiple processes) *but* one of those windows is minimized. 

My guess for fixing this is either:

- The tile is created before the window is minimized and the kwin script doesn't detect the window has been programatically minimized

- Or the kwin script responds to newly created windows but doesn't anticipate that one of them may be minimized.

**Comment by faho** [comment]
@TomBZombie: Please try 676b74d.

**Comment by TomBZombie** [comment]
That's fixed it for me. It probably explains why @piotr-dobrogost  was having the same problem with google chrome dev tools, the dev tools window was probably being started minimized and having a tile created.

**Comment by faho** [comment]
Okay, closing.

**Comment by TomBZombie** [comment]
Although that helped in that instance, it has happened again. Unfortunately, closing all windows and reopening them doesn't even fix it.

**Comment by TomBZombie** [comment]
For what it's worth, I've been using this for a couple of weeks now and never had this problem until adding a second monitor yesterday. I've also never had the issue on the primary display. Could it be something to do with the window being started initially on primary display?

**Comment by faho** [comment]
I absolutely cannot reproduce this. I've attached a second monitor, tried chromium's dev tools (pressing CTRL+SHIFT+D over and over again to attach/detach the dev tools window), tried x2goclient (which does not open a second tiled window for me), moved windows back and forth, changed the primary monitor.... There's something I'm missing, and I can't find it.

>Would it be a massive overhead to run an error-check during each window change? Iterate over each tile and check there's actually a window in it.

From some quick testing it seems noticeable when resizing. In contents/code/tiling.js, `Tiling.prototype._updateAllTiles`, add after `this.resizeScreen();`:

```js
for (var i = this.tiles.length - 1; i >= 0; i--) {
    if (this.tiles[i].clients.length == 0) {
        this.removeTile(this.tiles[i]);
    }
}
```

Then update the script. Plus, since I can't actually reproduce this I have no idea if it even works.

**Comment by TomBZombie** [comment]
I'm trying to replicate and debug this. It certainly happens fairly frequently. I've had it happen on two different virtual desktops this morning alone. Is there any way to display or even log to a file the contents of `this.tiles` each time it changes? I'd like to just keep track of exactly what's going on within the script.

**Comment by faho** [comment]
>Is there any way to display or even log to a file the contents of this.tiles each time it changes?

Log? No. The only way kwin scripts have to output anything is the `print` function, which will end up in kwin's console output. Or if you're using systemd, this'll end up in the journal, so `journalctl /usr/bin/kwin_x11 -b -f` will follow kwin's current output.

Note that this output is currently a bit sparse - we mostly print errors (since there's also no way to mark anything as debug output so it'll be shown with lower priority). I've toyed with the idea of adding a keybinding to e.g. mark a tile to be debugged. That would work well, but marking e.g. the layout logic or tilelist (the thing that reacts to clients starting and deciding whether they should be tiled or not) would require additional bindings.

>Alternatively, can kwin scripts read each others variables? Are there any global kwin script variables that could be set by kwin tiling and then read in wm console?

Honestly, I don't know. The wm console never really worked for me.

**Comment by TomBZombie** [comment]
Strangely I've not had the problem since I posted that. I rebooted, I can't imagine that makes any difference, after updating the script previously I logged out and back in but yesterday and this morning assuming that would have done it. And it did seem to fix the issue with sublime text

I will let you know if happens again and try to find a way of consistently replicating it.

**Comment by TomBZombie** [comment]
That post was 6 days ago, I haven't had it happen since then until now. I managed to trigger it twice but I can't seem to do it again! 

Here's what I did:

1. On screen 1 have Chromium (sole tile)
2. Two applications 50/50 on the secondary monitor
3. Drag a single tab out of Chromium onto the secondary display. (That's a single tab, creating a new window)

This caused two things:

1. The new window wasn't tiled, it was floating. As soon as I tried to move it, it was correctly positioned inside a tile

2. A phantom tile

It looks like the tile was created correctly but something prevented the window being moved into position. When I tried to manually move the window, a new tile was created for it.

**Comment by WildPenquin** [comment]
I have this problem too, but I have no idea what triggered it. There was no other way (I could find) to fix it except logging in and out again. I will see if I can find out how to reproduce and report back here, but this has definitely not been fixed.

**Comment by WildPenquin** [comment]
Ok, I figured out one way to reproduce this and possibly the only reason I've (personally) triggered this bug.

Windows will be kind of half-registered in the eyes of kwin-tiling if they existed when the script was enabled (possibly visiting kwin-tiling options could trigger this, if the script is re-loaded; I'm not sure how it handles this currently). Let me explain:

1) Enable kwin-tiling.
2) Open some (say, three, but even one should suffice) windows (konsole is a good test case, but it should not matter what program you use AFAICT). Optional: you can close and open as many windows and as many times as you like, and see that tiling works as expected, as long as you do not close (all of) the windows you have opened!
3) Disable kwin-tiling in system settings (remember to click "apply").
4) Re-enable kwin-tiling (again, click "apply" or exit settings).
5) Open more windows in a desktop you had opened some previously, things should work as expected (new window will be tiled among / with the old ones)
6) Close a window opened after re-enabling kwin-tiling. This still should work as expected.
7) Close a window opened **before** re-enabling kwin-tiling. Observe a gap / "phantom window"!

Also, you can "fix" the gaps by disabling and re-enabling the script; but you need to close all windows (or, expect the old ones leave phantom windows if / when you close them). As a bright side, this bug should not happen if you do not disable and re-enable (and possibly don't change settings) in kwin-tiling. This would also explain why this bug might happen for some and then disappear (since they would only experience this the very first time(s) they enable kwin-tiling while setting it up).

**Comment by TomBZombie** [comment]
I'm afraid it's not just that. I never reload the script and the bug happens periodically. It would be nice if there was a feature that allowed resetting, have the script scan all windows and re-tile them. At least that would be a better workaround than logging in and out as I have to currently.

**Comment by WildPenquin** [comment]
There certainly can be other causes.

However, IFAICT disabling and enabling "resets" the script; the problem here is that it does not properly register old windows, which means you still need to close all windows before re-enabling, which is certainly not optimal (you need to re-open evrything manually). In any case, I presume this needs to be fixed before it makes sense to implement a reset feature, since it could not possibly work properly in the current state.

I'm actually suprised no one else has noticed this, since this also means this bug will always trigger on the very first start of using the script (unless the user has no windows open previously; or logs out and back in immediately).

**Comment by faho** [comment]
@WildPenquin: Do you have 2f4b9c8676f731c370cb1701e0f0233a35b778e9? Because for quite a while, there was a kwin bug that prevented us from adding preexisting clients, and that only recently reenabled it.

So either you don't have that commit - in which case the script breaking with existing clients is unavoidable, or you _do_ have it and there's _something_ wrong.

Personally, I can't reproduce your problem. Which KWin version are you running?

>possibly visiting kwin-tiling options could trigger this, if the script is re-loaded; I'm not sure how it handles visiting settings currently

Nope,  the script gets absolutely no notification of that, or changing options (there's _supposed_ to be a signal for that, but AFAICT it never worked properly).

**Comment by WildPenquin** [comment]
> @WildPenquin: Do you have 2f4b9c8? Because for quite a while, there was a kwin bug that prevented us from adding preexisting clients, and that only recently reenabled it.

Yes, I have it; I'm actually on Arch and have installed from AUR [(kwin-scripts-tiling-git](https://aur.archlinux.org/packages/kwin-scripts-tiling-git/) which seems will always install from current HEAD; but just to confirm I checked what I have in /usr/share/kwin/scripts/kwin-script-tiling/contents/code/tilingmanager.js and the code is there).
```
$ kwin_x11 --version 
kwin 5.11.5
```

**Comment by faho** [comment]
Okay, with `kwin_x11 --replace` running in a terminal, I get

>tilelist.js:78: ReferenceError: options is not defined

After reenabling the script.

If I remove that use of "options", I get other fun stuff like

>tilelist.js:378: ReferenceError: KWin is not defined

>tilingmanager.js:812: ReferenceError: workspace is not defined

In other words: It's an upstream issue. Those are global variables that kwin is supposed to define for us, but it seems that when enabling a script that doesn't happen.

And in that case there really is nothing the script _can_ do. We can't have a "fix tiling" binding because to fix tiling we'd have to get the list of clients again, which is in workspace. To read the config, we need KWin. To figure out if compositing is enabled (which means we need to add clients differently) we need options.

**Comment by tflori** [comment]
I have phantom windows at least once a day and restarting kwin every time is a bit annoying. It might be complicated to figure out how it happens but it would be a plan to have a shortcut (maybe on a separate branch?) to iterate over all tiles on the current monitor, check if the client is still "valid" and if not log what client it was (wm_class, or some other properties) and remove the tile.

Because even if we don't get a list of clients we have a list of tiles and they have a property to their client. I mean we are not really interested in all clients but in clients of our tiles.

Btw. mostly I get this errors with phpstorm (I guess any Jetbrains IDE will work). I have disabled to open the previous project and therefore I get a list of projects first. This window cannot be resized but it get's a tile (it's like a floating window but the space of a tile is empty). Sometimes when I open a project from this window the tile will not go away but the window does. So with the "remove empty tiles" method I expect the wm_class would be `jetbrains_phpstorm` but I'm not sure.

**Comment by laloch** [comment]
@tflori: I have noticed that you do actually use the patch modified to remove tiles on your own fork/branch. Have you seen it to delete any empty tiles yet?

**Comment by tflori** [comment]
@laloch not yet. yesterday I tried to reproduce it again but it worked fine. this is such a bug that hides if a predator is near :slightly_smiling_face:

I will stay on this branch and report if something happens.. @faho thx for this feature

**Comment by faho** [comment]
Yeah, I have no idea how to reproduce this, so I'm closing. Sorry!

**Comment by 4n0nct** [comment]
No don't close please, there are still phantom windows and empty spaces

**Comment by faho** [comment]
@4n0nct This is closed because nobody could provide a way to reproduce it. I haven't seen it, and nobody has figured out when it happens.

So, because that makes it impossible to solve, it's closed. If someone figures out what causes it, we can reopen the issue or open a new one.

**Comment by rhopp** [comment]
The most common way how to reproduce this is to use multimonitor setup (I have 2 monitors) and turning one monitor off, so all the windows are moved to single monitor and then turning the monitor back on.
After this, the tiling is usually "broken" in the sense of phantom windows being present (even though no window/application disappeared. It's like the new phantom window was added).
(My system - Fedora 35, X11, Plasma 5.23.3)

**Comment by rhopp** [comment]
The most usual way for me how to reproduce is to turn off one (of two) monitor and then turn it on again. Even though no window was lost in the process, new phantom window appears in tiling.
My system (Fedora 35, X11, Plasma 5.23.3). 
Unfortunately the dump-clients shortcut is somehow not working for me (I don't see anything in journal after pressing it)
